crates/ethereum/src/lib.rs

```diff
@@ -1,111 +1,434 @@
+    //
+    // $ curl https://external.integration.starknet.io/feeder_gateway/get_contract_addresses
+    // {"GpsStatementVerifier": "0x8f97970aC5a9aa8D130d35146F5b59c4aef57963", "Starknet": "0xd5c325D183C592C94998000C5e0EED9e6655c020"}
+
+    pub const MAINNET: [u8; 20] = Decoder::Hex.decode(b"c662c410C0ECf747543f5bA90660f6ABeBD9C8c4");
+    pub const TESTNET: [u8; 20] = Decoder::Hex.decode(b"de29d060D45901Fb19ED6C6e959EB22d8626708e");
+    pub const TESTNET2: [u8; 20] = Decoder::Hex.decode(b"a4eD3aD27c294565cB0DCc993BDdCC75432D498c");
+    pub const INTEGRATION: [u8; 20] =
+        Decoder::Hex.decode(b"d5c325D183C592C94998000C5e0EED9e6655c020");
 }
 
-impl TryFrom<&ethers::types::Log> for BlockOrigin {
-    type Error = anyhow::Error;
+impl EthereumClient {
+    async fn get_block_number(&self) -> anyhow::Result<U256> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_blockNumber",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let block_number = res["result"]
+            .as_str()
+            .and_then(|txt| U256::from_str_radix(txt, 16).ok())
+            .ok_or(anyhow::anyhow!("Failed to get block number"))?;
+        Ok(block_number)
+    }
 
-    fn try_from(log: &ethers::types::Log) -> Result<Self, Self::Error> {
-        let hash = log.block_hash.context("missing block hash")?;
-        let hash = EthereumBlockHash(hash);
-        let number = log.block_number.context("missing block number")?.as_u64();
-        let number = EthereumBlockNumber(number);
-        Ok(Self { hash, number })
+    pub async fn gas_price(&self) -> anyhow::Result<U256> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_gasPrice",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let block_number = res["result"]
+            .as_str()
+            .and_then(|txt| U256::from_str_radix(txt, 16).ok())
+            .ok_or(anyhow::anyhow!("Failed to get gas price"))?;
+        Ok(block_number)
+    }
+
+    pub async fn chain_id(&self) -> anyhow::Result<EthereumChain> {
+        let req = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "eth_chainId",
+            "params": [],
+            "id" :0
+        });
+        let res: serde_json::Value = self
+            .http
+            .post(self.url.as_str())
+            .json(&req)
+            .send()
+            .await?
+            .json()
+            .await?;
+        let id = res["result"]
```

Comment:
TIL this does not panic if the field is missing, but instead returns `Value::Null`.